discovery-provider/alembic/versions/5bcbe23f6c70_user_track_collection_mat_views.py

```diff
@@ -0,0 +1,238 @@
+        -- join on subquery for track saves
+        LEFT OUTER JOIN (
+            SELECT
+                s.user_id as user_id,
+                count(s.user_id) as save_count
+            FROM
+                saves s
+            WHERE
+                s.is_current is True AND
+                s.save_type = 'track' AND
+                s.is_delete is False
+            GROUP BY s.user_id
+        ) user_track_save ON user_track_save.user_id = u.user_id
+        WHERE
+            u.is_current is True;
+
+        CREATE UNIQUE INDEX aggregate_user_idx ON aggregate_user (user_id);    
+
+        --- ======================= AGGREGATE TRACK =======================
+        DROP MATERIALIZED VIEW IF EXISTS aggregate_track;
+        DROP INDEX IF EXISTS aggregate_track_idx;
+
+        CREATE MATERIALIZED VIEW aggregate_track as
+        SELECT
+          t.track_id,
+          COALESCE (track_repost.repost_count, 0) as repost_count,
+          COALESCE (track_save.save_count, 0) as save_count
+        FROM 
+          tracks t
+        -- inner join on subquery for reposts
+        LEFT OUTER JOIN (
+          SELECT
+            r.repost_item_id as track_id,
+            count(r.repost_item_id) as repost_count
+          FROM
+            reposts r
+          WHERE
+            r.is_current is True AND
+            r.repost_type = 'track' AND
+            r.is_delete is False
+          GROUP BY r.repost_item_id
+        ) track_repost ON track_repost.track_id = t.track_id
+        -- inner join on subquery for track saves
+        LEFT OUTER JOIN (
+          SELECT
+            s.save_item_id as track_id,
+            count(s.save_item_id) as save_count
+          FROM
+            saves s
+          WHERE
+            s.is_current is True AND
+            s.save_type = 'track' AND
+            s.is_delete is False
+          GROUP BY s.save_item_id
+        ) track_save ON track_save.track_id = t.track_id
+        WHERE
+          t.is_current is True AND
+          t.is_delete is False;
+
+        CREATE UNIQUE INDEX aggregate_track_idx ON aggregate_track (track_id);
+
+        --- ======================= AGGREGATE PLAYLIST =======================
+        DROP MATERIALIZED VIEW IF EXISTS aggregate_playlist;
+        DROP INDEX IF EXISTS aggregate_playlist_idx;
+
+        CREATE MATERIALIZED VIEW aggregate_playlist as
+        SELECT
+          p.playlist_id,
+          p.is_album,
+          COALESCE (playlist_repost.repost_count, 0) as repost_count,
+          COALESCE (playlist_save.save_count, 0) as save_count
+        FROM 
+          playlists p
+        -- inner join on subquery for reposts
+        LEFT OUTER JOIN (
+          SELECT
+            r.repost_item_id as playlist_id,
+            count(r.repost_item_id) as repost_count
+          FROM
+            reposts r
+          WHERE
+            r.is_current is True AND
+            (r.repost_type = 'playlist' OR r.repost_type = 'album') AND
```

Comment:
Does this work? If a playlist + album have the same ID, won't this count both for both?

Reply:
an album is technically a playlist under the hood, just with the `is_album` bool set to true, so there should not be interesting playlist_ids between the two - which I took advantage of to write this in a simpler query.

Reply:
Right ofc